### ISSUE TITLE
Modifier to add time dimension

### ISSUE THREAD

**Issue by gerritholl** [pytroll/satpy#2428]
## Feature Request

**Is your feature request related to a problem? Please describe.**

We have several tools related to nowcasting that require input data to have a singleton time dimension, even if the data are relating to a single time step and produced from input data that do not have a time dimension.  For example, when reading a NetCDF file, they expect variables to have dimensions `(time, y, x)` (where `time` has length 1) rather than just `(y, x)`.  When we read a single Scene for data that do not have a time dimension and then write a NetCDF file, the dimensions are (of course) `(y, x)`, as they should be.  To write `(time, y, x)`, we would need to modify the data first.

It is said that a Scene in Satpy relates to a single time step, but this is not entirely true.  With `MultiScene.blend(blend_function=timeseries)`, we can get a Scene that relates to multiple timesteps (and would be able to write this to NetCDF if not for https://github.com/pytroll/satpy/issues/2426).

**Describe the solution you'd like**

I would like a modifier (or maybe an enhancement?) that adds a singleton time-dimension, thus changing datasets from dimensions `(y, x)` to `(time, y, x)`.  The length-1 time dimension should have a coordinate with an appropriate reference time; probably `start_time`, but this could be configurable.

**Describe any changes to existing user workflow**

As new functionality, this does not affect user workflow.

**Additional context**

It would be possible to make a `MultiScene` that only consists of one Scene, then blend this with the timeseries blend to get a scene again.  That feels somewhat hackish and perhaps not entirely stable.  It's also hard to integrate into trollflow2 as it is at present, whereas a modifier or enhancement could be integrated into existing setups relatively easily.

**Comment by djhoese** [comment]
I would consider the `timeseries` blending functionality an exception to the "single time" rule. For example, saving that Scene to a geotiff wouldn't work because the geotiff writer doesn't know how to handle a time dimension. It is easier to describe the handling of time to a new user working with Scene creation and `.load` that "a Scene is meant for one time step". I'd say that's where that "rule" comes from.

As for this functionality, you primarily need this for writing a NetCDF file right? Isn't this functionality already supported by the CF writer? I can see this being nice to have outside of the CF writer, but just wanted to mention that it may be simple to add it to the CF writer. Anyway, I don't think this should be a modifier or an enhancement. This sounds like it should be a utility function. Modifiers and enhancements are generally meant to change the image data and then update the metadata to still be valid with that new data (single band compositor is an exception). This new functionality would be purely metadata.

**Comment by gerritholl** [comment]
Yes, we need this functionality only for writing a NetCDF file.  I don't think writing a time dimension to any other file format would be possible or suitable.

I think the CF writer can write time information, but my first attempts to do so have failed (https://github.com/pytroll/satpy/issues/2426), so I'm not sure what it actually supports.  In any case, it only writes time information if this information is in the input.

The writer feels like the wrong place to add information.  On the other hand, we also apply enhancers only in a `save_datesets` call, so maybe a utility function would still be better than an enhancement or similar?  Something like

`sc.save_datasets(writer="netcdf", writer_args={"write_time_dimension": "start_time"})` or such?

Or more generically?

`sc.save_datasets(writer="netcdf", writer_args={"pre_function": (satpy.writers.CFWriter.add_time_dimension, "start_time")})`

It's true that users writing NetCDF probably have somewhat different needs than users writing GeoTIFF.  They might not want to change anything to the data themselves, but may want more control over how (metadata) are written.  Of course a user with own Python code has full control, but in operations I like to build everything on existing building blocks from trollflow2 with local YAML config but without local scripts.

**Comment by mraspaud** [comment]
As to the usage of the time dimension: I would say that if a dataarray has a time dimension, and is being saved for imagery, I would assume an animation is created... just my 2 öre

**Comment by djhoese** [comment]
>  just my 2 öre

What's the conversion rate on that? When people say "2 cents" is that US?

> I would say that if a dataarray has a time dimension, and is being saved for imagery, I would assume an animation is created

I would agree with that, but writers are currently not animation compatible. I'm not sure they should be...unless we want to add a `save_animation` to `Scene` that mirrors/uses `MultiScene.save_animation`. Oh...that's an interesting code cleanup. We would have `MultiScene.save_animation` use `.blend(timeseries)` and then `Scene.save_animation`...there probably aren't as many possible optimizations with that though.

**Comment by gerritholl** [comment]
> > just my 2 öre
> 
> What's the conversion rate on that? When people say "2 cents" is that US?

2 öre is around 0.19 US cents, so it's an even more modest contribution ☺

I think the idea that an animation would be saved is interesting, but I think that would only work with `simple_image`.  For TIFF, some users might expect that each frame would become a different band, but that would only work for modes L or P.  But we could also save each timestep in a different file.

But the idea to have a time dimension of length one might be more unique to NetCDF files that are passed on to nowcasting utilities that expect this form of input data (I think cdo is such a tool).

**Comment by mraspaud** [comment]
Thinking more about this, I suppose if you save an image with multiple time steps to a tiff or png, you should either get an image for each time step or a blending of the data. And for animations, that would assume we have winter for avi or mp4 files.

**Comment by gerritholl** [comment]
For PNG a user might expect APNG, but perhaps this would be configurable.

**Comment by jdaniel006** [comment]
Hey! I second this suggestion, as netCDFs with time dimensions would be greatly useful.

Trying to manually add the time dimension to the datasets (with assign _coords and expand_dims) hasn't worked so far. cf_writer always says "No time dimension in datasets". Maybe I'm doing something wrong. The goal is to set time to the start_time of a SEVIRI native file.

**Comment by mraspaud** [comment]
@jdaniel006 could you show us how you are doing?

**Comment by jdaniel006** [comment]
In a quick and dirty way, after creating a Scene and loading channels:
```
# Get start date
scene_date= compScene.start_time.strftime("%Y%m%d%H%M")
# Get channel ids
allChannels = compScene.available_dataset_names()
# Convert date 
dt = pd.to_datetime(scene_date)
# Add time to each channel
for channel in allChannels:
    compScene[channel]=compScene[channel].assign_coords(time=dt)
    compScene[channel]=compScene[channel].expand_dims(dim={"time":1})
```

With that, the scene can be saved with time dimension. Maybe not CF compliant and very likely the double conversion of start_time is not the most clever thing, but it has worked.
